pkg/autoscaler/autoscaler.go

```diff
@@ -434,4 +446,72 @@ var _ = Describe("[Feature:Machines] Autoscaler should", func() {
 			framework.WaitForMachineSetDelete(client, ms)
 		}
 	})
+
+	It("It scales from zero", func() {
+		defer GinkgoRecover()
+
+		// Only run in platforms which support autoscaling from/to zero.
+		clusterInfra, err := framework.GetInfrastructure(client)
+		Expect(err).NotTo(HaveOccurred())
+
+		platform := clusterInfra.Status.PlatformStatus.Type
+		switch platform {
+		case configv1.AWSPlatformType, configv1.GCPPlatformType, configv1.AzurePlatformType:
+			glog.Infof("Platform is %v", platform)
+		default:
+			Skip(fmt.Sprintf("Platform %v does not support autoscaling from/to zero, skipping.", platform))
+		}
+
+		// Anything we create we must cleanup
+		cleanupObjects := map[string]runtime.Object{}
+		defer func() {
+			for name, obj := range cleanupObjects {
+				if err := deleteObject(name, obj); err != nil {
+					glog.Infof("[cleanup] error deleting object %q (%T): %v", name, obj, err)
+				}
+			}
+		}()
+
+		By("Creating a new MachineSet with 0 replicas")
+		machineSetParams := framework.BuildMachineSetParams(client, 0)
+		machineSetParams.Labels[autoscalerWorkerNodeRoleLabel] = ""
+
+		machineSet, err := framework.CreateMachineSet(client, machineSetParams)
+		Expect(err).ToNot(HaveOccurred())
+		cleanupObjects[machineSet.GetName()] = runtime.Object(machineSet)
+
+		framework.WaitForMachineSet(client, machineSet.GetName())
+
+		By(fmt.Sprintf("Creating ClusterAutoscaler "))
+		clusterAutoscaler := clusterAutoscalerResource(100)
+		Expect(client.Create(context.TODO(), clusterAutoscaler)).Should(Succeed())
```

Comment:
Might be worth setting a variable at the top of this describe `ctx = context.Background()` and using that instead of `TODO()` everywhere?